File: mahjong/src/main/java/com/code/server/game/mahjong/service/MahjongRoomService.java
package com.code.server.game.mahjong.service;

import com.code.server.constant.game.IGameConstant;
import com.code.server.constant.response.ResponseVo;
import com.code.server.game.mahjong.config.ServerConfig;
import com.code.server.game.mahjong.logic.RoomFactory;
import com.code.server.game.mahjong.logic.RoomInfo;
import com.code.server.game.mahjong.util.ErrorCode;
import com.code.server.game.mahjong.util.SpringUtil;
import com.code.server.game.room.MsgSender;
import com.code.server.game.room.Room;
import com.code.server.game.room.service.RoomManager;
import com.code.server.redis.service.RedisManager;
import com.code.server.util.timer.GameTimer;
import com.code.server.util.timer.ITimeHandler;
import com.code.server.util.timer.TimerNode;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Created by sunxianping on 2017/6/5.
 */
public class MahjongRoomService {

    public static int dispatch(long userId, String method, JsonNode paramsjSONObject) {
        int code = 0;
        switch (method) {
            case "createRoomByUser": {
                String modeTotal = paramsjSONObject.get("modeTotal").asText();
                String mode = paramsjSONObject.get("mode").asText();
                int multiple = paramsjSONObject.get("multiple").asInt();
                int gameNumber = paramsjSONObject.get("gameNumber").asInt();
                int personNumber = paramsjSONObject.get("personNumber").asInt();
                String gameType = paramsjSONObject.get("gameType").asText();
                code = createRoomByUser(userId, modeTotal, mode, multiple, gameNumber, personNumber, gameType);
                break;
            }
            case "createRoomByEachUser": {
                String modeTotal = paramsjSONObject.get("modeTotal").asText();
                String mode = paramsjSONObject.get("mode").asText();
                int multiple = paramsjSONObject.get("multiple").asInt();
                int gameNumber = paramsjSONObject.get("gameNumber").asInt();
                int personNumber = paramsjSONObject.get("personNumber").asInt();
                String gameType = paramsjSONObject.get("gameType").asText();
                code = createRoomByEachUser(userId, modeTotal, mode, multiple, gameNumber, personNumber, gameType);
                break;
            }

            case "createRoomButNotInRoom": {
                String modeTotal = paramsjSONObject.get("modeTotal").asText();
                String mode = paramsjSONObject.get("mode").asText();
                int multiple = paramsjSONObject.get("multiple").asInt();
                int gameNumber = paramsjSONObject.get("gameNumber").asInt();
                int personNumber = paramsjSONObject.get("personNumber").asInt();
                String gameType = paramsjSONObject.get("gameType").asText();
                code = createRoomButNotInRoom(userId, modeTotal, mode, multiple, gameNumber, personNumber, gameType);
                break;
            }
        }


        return  code;
    }






    private static boolean isCanCreate(String modeTotal, String mode, String multiple) {
        if (modeTotal.equals("1")) {
            if (!mode.equals("5") && !mode.equals("6")) {
                return false;
            }
            if (!multiple.equals("1") && !multiple.equals("2") && !multiple.equals("5")) {
                return false;
            }
        } else if (modeTotal.equals("2")) {
            if (!mode.equals("1") && !mode.equals("2") && !mode.equals("3") && !mode.equals("4")) {
                return false;
            }
            if (!multiple.equals("1") && !multiple.equals("2") && !multiple.equals("5")) {
                return false;
            }
        } else if (modeTotal.equals("3") || modeTotal.equals("15")) {
            if (!multiple.equals("1") && !multiple.equals("2") && !multiple.equals("5")) {
                return false;
            }
        } else if (modeTotal.equals("4")) {
            if (!mode.equals("303")) {
                return false;
            }
        } else if (modeTotal.equals("5")) {
            if (!multiple.equals("1") && !multiple.equals("2") && !multiple.equals("5")) {
                return false;
            }
        } else if (modeTotal.equals("6")) {
            if (!mode.equals("0")) {
                return false;
            }
        }
        /**
         256封顶、128封顶、杠呲宝、未上听包三家、杠上开花、三七夹	二进制取,1是0否
         */
        else if (modeTotal.equals("10")) {
            if (!multiple.equals("25") && !multiple.equals("50") && !multiple.equals("100") && !multiple.equals("200")) {
                return false;
            }
            if (Integer.parseInt(mode) > 63 || Integer.parseInt(mode) < 0) {
                return false;
            }
        } else if (modeTotal.equals("11")) {
            if (!multiple.equals("25") && !multiple.equals("50") && !multiple.equals("100") && !multiple.equals("200")) {
                return false;
            }
            if (Integer.parseInt(mode) > 63 || Integer.parseInt(mode) < 0) {
                return false;
            }
        } else if (modeTotal.equals("12")) {
            if (!multiple.equals("1") && !multiple.equals("2")) {
                return false;
            }
            if (Integer.parseInt(mode) != 12051314 && Integer.parseInt(mode) != 12051304 && Integer.parseInt(mode) != 12050314 && Integer.parseInt(mode) != 12050304 && Integer.parseInt(mode) != 12151314 && Integer.parseInt(mode) != 12151304 && Integer.parseInt(mode) != 12150314 && Integer.parseInt(mode) != 12150304) {
                return false;
            }
        } else if (modeTotal.equals("13")) {
            if (Integer.parseInt(mode) != 13051314 && Integer.parseInt(mode) != 13051304 && Integer.parseInt(mode) != 13050314 && Integer.parseInt(mode) != 13050304 && Integer.parseInt(mode) != 13151314 && Integer.parseInt(mode) != 13151304 && Integer.parseInt(mode) != 13150314 && Integer.parseInt(mode) != 13150304) {
                return false;
            }
        } else if (modeTotal.equals("14")) {
            if (!multiple.equals("25") && !multiple.equals("50") && !multiple.equals("100") && !multiple.equals("200")) {
                return false;
            }
            if (Integer.parseInt(mode) > 63 || Integer.parseInt(mode) < 0) {
                return false;
            }
        } else if (modeTotal.equals("124")) {
            if (Integer.parseInt(mode) != 124051314 && Integer.parseInt(mode) != 124051304 && Integer.parseInt(mode) != 124050314 && Integer.parseInt(mode) != 124050304 && Integer.parseInt(mode) != 124151314 && Integer.parseInt(mode) != 124151304 && Integer.parseInt(mode) != 124150314 && Integer.parseInt(mode) != 124150304) {
                return false;
            }
        } else {
            return false;
        }
        return true;
    }

    public static RoomInfo createRoom(long userId, String modeTotal, String mode, int multiple, int gameNumber, int personNumber, String gameType, String each, boolean isJoin) {
        if (!isCanCreate(modeTotal, mode, "" + multiple)) {
            return null;
        }

        RoomInfo roomInfo = RoomFactory.getRoomInstance(gameType);
        String roomId = Room.getRoomIdStr(Room.genRoomId());
        roomInfo.init(roomId, userId, modeTotal, mode, multiple, gameNumber, personNumber, userId, 0);
        roomInfo.setEach(each);
        if (isJoin) {

            roomInfo.joinRoom(userId);
        }

        int serverId = SpringUtil.getBean(ServerConfig.class).getServerId();
        RoomManager.addRoom(roomInfo.getRoomId(), "" + serverId, roomInfo);

        return roomInfo;
    }


    public static int createRoomByEachUser(long userId, String modeTotal, String mode, int multiple, int gameNumber, int personNumber, String gameType) {
        RoomInfo roomInfo = createRoom(userId, modeTotal, mode, multiple, gameNumber, personNumber, gameType, "1", true);
        if (roomInfo == null) {
            return ErrorCode.CANNOT_CREATE_ROOM_PARAMETER_IS_ERROR;
        }
        MsgSender.sendMsg2Player(new ResponseVo("mahjongRoomService", "createRoomByEachUser", roomInfo.toJSONObject()), userId);
        return 0;
    }

    public static int createRoomByUser(long userId, String modeTotal, String mode, int multiple, int gameNumber, int personNumber, String gameType) {
        RoomInfo roomInfo = createRoom(userId, modeTotal, mode, multiple, gameNumber, personNumber, gameType, "0", true);
        if (roomInfo == null) {
            return ErrorCode.CANNOT_CREATE_ROOM_PARAMETER_IS_ERROR;
        }
        MsgSender.sendMsg2Player(new ResponseVo("mahjongRoomService", "createRoomByUser", roomInfo.toJSONObject()), userId);
        return 0;
    }

    public static int createRoomButNotInRoom(long userId, String modeTotal, String mode, int multiple, int gameNumber, int personNumber, String gameType) {

        if ("LQ".equals(gameType)) {
            double money = RedisManager.getUserRedisService().getUserMoney(userId);
            if (8 == gameNumber) {
                if (money < 30) {
                    return ErrorCode.CANNOT_JOIN_ROOM_NO_MONEY;
                }
                RedisManager.getUserRedisService().addUserMoney(userId, -30);
                RedisManager.addGold(userId, 3);

            } else if (16 == gameNumber) {
                if (money < 60) {
                    return ErrorCode.CANNOT_JOIN_ROOM_NO_MONEY;
                }
                RedisManager.getUserRedisService().addUserMoney(userId, -60);
                RedisManager.addGold(userId, 6);
            }
        }

        RoomInfo roomInfo = createRoom(userId, modeTotal, mode, multiple, gameNumber, personNumber, gameType, "2", false);

        if (roomInfo == null) {
            return ErrorCode.CANNOT_CREATE_ROOM_PARAMETER_IS_ERROR;
        }

        String roomId = roomInfo.getRoomId();


        long start = System.currentTimeMillis();
        TimerNode node = new TimerNode(start, IGameConstant.ONE_HOUR, false, new ITimeHandler() {
            @Override
            public void fire() {
                try {
                    RoomInfo roomInfo = (RoomInfo) RoomManager.getRoom(roomId);

                    if (roomInfo != null && !roomInfo.isInGame() && roomInfo.getCurGameNumber() == 1) {

                        if ("LQ".equals(roomInfo.getGameType())) {
                            if (8 == roomInfo.getGameNumber()) {

                                RedisManager.getUserRedisService().addUserMoney(userId, 30);
                                RedisManager.addGold(userId, -3);
                            } else if (16 == roomInfo.getGameNumber()) {
                                RedisManager.getUserRedisService().addUserMoney(userId, 60);
                                RedisManager.addGold(userId, -6);
                            }
                        }
                        RoomManager.removeRoom(roomInfo.getRoomId());
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
        roomInfo.setTimerNode(node);
        GameTimer.getInstance().addTimerNode(node);


        MsgSender.sendMsg2Player(new ResponseVo("mahjongRoomService", "createRoomButNotInRoom", roomInfo.toJSONObject()), userId);
        return 0;
    }


//    public void onlinemethod(GamePlayer gamePlayer) {
//
//        Map<Integer, String> userRoom = GameManager.getInstance().getUserRoom();
//
//        String roomId = userRoom.get(gamePlayer.getUserId());//房间号
//
//        RoomInfo roomInfo = GameManager.getInstance().getRoom(roomId);
//
//        if (roomInfo != null) {
//            List<Integer> roomUserId = roomInfo.getUsers();
//
//            JSONObject result = new JSONObject();
//
//            Map<String, Integer> mapresult = new HashMap<String, Integer>();
//
//            mapresult.put("id", gamePlayer.getUserId());
//            mapresult.put("status", 1);
//
//            result.put("service", "gameService");
//            result.put("method", "offline");
//            result.put("params", new Gson().toJson(mapresult));
//            result.put("code", "0");
//
//            serverContext.sendToOnlinePlayer(result, roomUserId);
//        }
//    }
}

File: mahjong/src/main/scala/com/code/server/game/mahjong/service/MahjongRoomService.scala
package com.code.server.game.mahjong.service

import com.code.server.constant.response.{ErrorCode, ResponseVo}
import com.code.server.game.mahjong.config.ServerConfig
import com.code.server.game.mahjong.logic.{RoomFactory, RoomInfo, RoomInfoExtendGold}
import com.code.server.game.room.{Room, RoomExtendGold}
import com.code.server.game.room.kafka.MsgSender
import com.code.server.game.room.service.RoomManager
import com.code.server.redis.service.RedisManager
import com.code.server.util.timer.{GameTimer, TimerNode}
import com.code.server.util.{IdWorker, SpringUtil}
import com.fasterxml.jackson.databind.JsonNode

/**
  * Created by sunxianping on 2017/6/5.
  */
object MahjongRoomService {
  def dispatch(userId: Long, method: String, paramsjSONObject: JsonNode): Int = {
    var code: Int = 0
    method match {
      case "createRoomByUser" => {
        val modeTotal: String = paramsjSONObject.get("modeTotal").asText
        val mode: String = paramsjSONObject.get("mode").asText
        val multiple: Int = paramsjSONObject.get("multiple").asInt
        val gameNumber: Int = paramsjSONObject.get("gameNumber").asInt
        val personNumber: Int = paramsjSONObject.get("personNumber").asInt
        val gameType: String = paramsjSONObject.get("gameType").asText
        val roomType = paramsjSONObject.get("roomType").asText
        val mustZimo = paramsjSONObject.path("mustZimo").asInt
        val isHasYipaoduoxiang = paramsjSONObject.path("yipaoduoxiang").asBoolean(false)
        val canChi = paramsjSONObject.path("canChi").asBoolean(false)
        val haveTing = paramsjSONObject.path("haveTing").asBoolean(false)
        val clubId = paramsjSONObject.path("clubId").asText
        val clubRoomModel = paramsjSONObject.path("clubRoomModel").asText
        val clubMode = paramsjSONObject.path("clubMode").asInt(0)
        val goldRoomType = paramsjSONObject.path("goldRoomType").asInt(0)
        val goldRoomPermission = paramsjSONObject.path("goldRoomPermission").asInt(0)
        val showChat = paramsjSONObject.path("showChat").asBoolean(true)
        val otherMode = paramsjSONObject.path("otherMode").asInt(0)

        code = createRoomByUser(userId, modeTotal, mode, multiple, gameNumber, personNumber, gameType,
          roomType, mustZimo, isHasYipaoduoxiang, canChi, haveTing,
          clubId, clubRoomModel, clubMode, goldRoomType, goldRoomPermission, showChat, otherMode)

      }
      case "createRoomByEachUser" => {
        val modeTotal: String = paramsjSONObject.get("modeTotal").asText
        val mode: String = paramsjSONObject.get("mode").asText
        val multiple: Int = paramsjSONObject.get("multiple").asInt
        val gameNumber: Int = paramsjSONObject.get("gameNumber").asInt
        val personNumber: Int = paramsjSONObject.get("personNumber").asInt
        val gameType: String = paramsjSONObject.get("gameType").asText
        val roomType = paramsjSONObject.get("roomType").asText()
        val mustZimo = paramsjSONObject.path("mustZimo").asInt
        val isHasYipaoduoxiang = paramsjSONObject.path("yipaoduoxiang").asBoolean(false)
        val canChi = paramsjSONObject.path("canChi").asBoolean(false)
        val haveTing = paramsjSONObject.path("haveTing").asBoolean(false)
        val clubId = paramsjSONObject.path("clubId").asText
        val clubRoomModel = paramsjSONObject.path("clubRoomModel").asText
        val clubMode = paramsjSONObject.path("clubMode").asInt(0)
        val goldRoomType = paramsjSONObject.path("goldRoomType").asInt(0)
        val goldRoomPermission = paramsjSONObject.path("goldRoomPermission").asInt(0)
        val showChat = paramsjSONObject.path("showChat").asBoolean(true)
        val otherMode = paramsjSONObject.path("otherMode").asInt(0)
        code = createRoomByEachUser(userId, modeTotal, mode, multiple, gameNumber,
          personNumber, gameType, roomType, mustZimo, isHasYipaoduoxiang, canChi, haveTing,
          clubId, clubRoomModel, clubMode, goldRoomType, goldRoomPermission, showChat, otherMode)

      }
      case "createRoomButNotInRoom" => {
        val modeTotal: String = paramsjSONObject.get("modeTotal").asText
        val mode: String = paramsjSONObject.get("mode").asText
        val multiple: Int = paramsjSONObject.get("multiple").asInt
        val gameNumber: Int = paramsjSONObject.get("gameNumber").asInt
        val personNumber: Int = paramsjSONObject.get("personNumber").asInt
        val gameType: String = paramsjSONObject.get("gameType").asText
        val roomType = paramsjSONObject.get("roomType").asText()
        val mustZimo: Int = paramsjSONObject.path("mustZimo").asInt
        val isHasYipaoduoxiang = paramsjSONObject.path("yipaoduoxiang").asBoolean(false)
        val canChi = paramsjSONObject.path("canChi").asBoolean(false)
        val haveTing = paramsjSONObject.path("haveTing").asBoolean(false)
        val clubId = paramsjSONObject.path("clubId").asText
        val clubRoomModel = paramsjSONObject.path("clubRoomModel").asText
        val clubMode = paramsjSONObject.path("clubMode").asInt(0)
        val goldRoomType = paramsjSONObject.path("goldRoomType").asInt(0)
        val goldRoomPermission = paramsjSONObject.path("goldRoomPermission").asInt(0)
        val showChat = paramsjSONObject.path("showChat").asBoolean(true)
        val otherMode = paramsjSONObject.path("otherMode").asInt(0)
        code = createRoomButNotInRoom(userId, modeTotal, mode, multiple, gameNumber, personNumber,
          gameType, roomType, mustZimo, isHasYipaoduoxiang, canChi, haveTing,
          clubId, clubRoomModel, clubMode, goldRoomType, goldRoomPermission, showChat, otherMode)

      }

      case "joinGoldRoom" => {

        val roomType: String = paramsjSONObject.get("roomType").asText
        val gameType: String = paramsjSONObject.get("gameType").asText
        val goldRoomType = paramsjSONObject.path("goldRoomType").asInt()

        code = joinGoldRoom(userId, roomType, gameType, goldRoomType)

      }

      case "getGoldRooms" => {
        //        val roomType: String = paramsjSONObject.get("roomType").asText
        val gameType: String = paramsjSONObject.get("gameType").asText
        //        val goldRoomType = paramsjSONObject.path("goldRoomType").asInt(0)

        val result = RoomExtendGold.getGoldRoomsVo(gameType)
        MsgSender.sendMsg2Player("mahjongRoomService", "getGoldRooms", result, userId)
        code = 0
      }


      case "getRoomGameNum" => {
        val rooms = paramsjSONObject.path("rooms").asText()
        MsgSender.sendMsg2Player("mahjongRoomService", "getRoomGameNum", Room.getRoomNum(rooms), userId)
        0
      }


    }
    return code
  }

  private def isCanCreate(modeTotal: String, mode: String, multiple: String): Boolean = {
    if (modeTotal == "1") {
      if (mode != "5" && mode != "6") {
        return false
      }
      if (multiple != "1" && multiple != "2" && multiple != "5") {
        return false
      }
    }
    else if (modeTotal == "2") {
      if (mode != "1" && mode != "2" && mode != "3" && mode != "4" && mode != "11" && mode != "12" && mode != "13" && mode != "14") {
        return false
      }
      if (multiple != "1" && multiple != "2" && multiple != "3" && multiple != "5") {
        return false
      }
    }
    else if (modeTotal == "3" || modeTotal == "15") {
      if (multiple != "1" && multiple != "2" && multiple != "5") {
        return false
      }
    }
    else if (modeTotal == "4") {
      if (mode != "303") {
        return false
      }
    }
    else if (modeTotal == "5") {
      if (multiple != "1" && multiple != "2" && multiple != "5") {
        return false
      }
    }
    else if (modeTotal == "6") {
      if (mode != "0") {
        return false
      }
    }
    else

    /**
      * 256封顶、128封顶、杠呲宝、未上听包三家、杠上开花、三七夹	二进制取,1是0否
      */ if (modeTotal == "10") {
      if (multiple != "25" && multiple != "50" && multiple != "100" && multiple != "200") {
        return false
      }
      if (mode.toInt > 63 || mode.toInt < 0) {
        return false
      }
    }
    else if (modeTotal == "11") {
      if (multiple != "25" && multiple != "50" && multiple != "100" && multiple != "200") {
        return false
      }
      if (mode.toInt > 63 || mode.toInt < 0) {
        return false
      }
    }
    else if (modeTotal == "12") {
      if (multiple != "1" && multiple != "2") {
        return false
      }
      if (mode.toInt != 12051314 && mode.toInt != 12051304 && mode.toInt != 12050314 && mode.toInt != 12050304 && mode.toInt != 12151314 && mode.toInt != 12151304 && mode.toInt != 12150314 && mode.toInt != 12150304) {
        return false
      }
    }
    else if (modeTotal == "13") {
      if (mode.toInt != 13051314 && mode.toInt != 13051304 && mode.toInt != 13050314 && mode.toInt != 13050304 && mode.toInt != 13151314 && mode.toInt != 13151304 && mode.toInt != 13150314 && mode.toInt != 13150304) {
        return false
      }
    }
    else if (modeTotal == "14") {
      if (multiple != "25" && multiple != "50" && multiple != "100" && multiple != "200") {
        return false
      }
      if (mode.toInt > 63 || mode.toInt < 0) {
        return false
      }
    }
    else if (modeTotal == "124") {
      if (mode.toInt != 124051314 && mode.toInt != 124051304 && mode.toInt != 124050314 && mode.toInt != 124050304 && mode.toInt != 124151314 && mode.toInt != 124151304 && mode.toInt != 124150314 && mode.toInt != 124150304) {
        return false
      }
    }
    else if (modeTotal == "20") {
      return true
    }
    else if (modeTotal == "30") {
      return true
    }
    else if (modeTotal == "31") {
      return true
    }
    else if (modeTotal == "100") {
      return true
    }
    else if (modeTotal == "33") {
      return true
    }
    else {
      return true
    }
    return true
  }

  def createRoom(userId: Long, modeTotal: String, mode: String, multiple: Int, gameNumber: Int, personNumber: Int, gameType: String,
                 each: String, isJoin: Boolean, roomType: String, mustZimo: Int, yipaoduoxiang: Boolean, canChi: Boolean, haveTing: Boolean,
                 clubId: String, clubRoomModel: String, clubMode: Int, goldRoomType: Int, goldRoomPermission: Int, showChat: Boolean, otherMode: Int): (Int, RoomInfo) = {
    val roomInfo: RoomInfo = RoomFactory.getRoomInstance(gameType)
    val serverId: Int = SpringUtil.getBean(classOf[ServerConfig]).getServerId
    val roomId: String = Room.getRoomIdStr(Room.genRoomId(serverId))
    roomInfo.setRoomType(roomType)
    roomInfo.setAA("1".equals(each))
    roomInfo.setEach(each)
    roomInfo.setClubId(clubId)
    roomInfo.setClubRoomModel(clubRoomModel)
    roomInfo.setCreaterJoin(isJoin)
    roomInfo.setYipaoduoxiang(yipaoduoxiang)
    roomInfo.setCanChi(canChi)
    roomInfo.setHaveTing(haveTing)
    roomInfo.setGoldRoomType(goldRoomType)
    roomInfo.setGoldRoomPermission(goldRoomPermission)
    roomInfo.setShowChat(showChat)
    roomInfo.setOtherMode(otherMode)
    roomInfo.setClubMode(clubMode)

    if (Room.isHasMode(2, otherMode)) {
      roomInfo.setRobotRoom(true)
    }
    roomInfo.init(roomId, userId, modeTotal, mode, multiple, gameNumber, personNumber, userId, 0, mustZimo)
    if (goldRoomPermission != 0) {
      roomInfo.setMultiple(goldRoomType)
    }
    var code = 0
    if (isJoin) {
      //todo 选座位
      code = roomInfo.joinRoom(userId, true)
      if (code != 0) {
        return (code, null)
      }
    }

    RoomManager.addRoom(roomInfo.getRoomId, "" + serverId, roomInfo)
    val idWorker = new IdWorker(serverId, 0)
    roomInfo.setUuid(idWorker.nextId())
    return (code, roomInfo)
  }

  /**
    * 四人建房
    *
    * @param userId
    * @param modeTotal
    * @param mode
    * @param multiple
    * @param gameNumber
    * @param personNumber
    * @param gameType
    * @param roomType
    * @param mustZimo
    * @param yipaoduoxiang
    * @param canChi
    * @param haveTing
    * @return
    */
  def createRoomByEachUser(userId: Long, modeTotal: String, mode: String, multiple: Int, gameNumber: Int, personNumber: Int,
                           gameType: String, roomType: String, mustZimo: Int, yipaoduoxiang: Boolean, canChi: Boolean, haveTing: Boolean,
                           clubId: String, clubRoomModel: String, clubMode: Int, goldRoomType: Int, goldRoomPermission: Int, showChat: Boolean, otherMode: Int): Int = {
    if (!isCanCreate(modeTotal, mode, "" + multiple)) {
      return ErrorCode.CANNOT_CREATE_ROOM_PARAMETER_IS_ERROR
    }
    val (code, roomInfo) = createRoom(userId, modeTotal, mode, multiple, gameNumber, personNumber, gameType, "1",
      true, roomType, mustZimo, yipaoduoxiang, canChi, haveTing,
      clubId, clubRoomModel, clubMode, goldRoomType, goldRoomPermission, showChat, otherMode)
    if (code != 0) {
      return code
    }

    if (roomInfo == null) {
      return ErrorCode.CANNOT_CREATE_ROOM_PARAMETER_IS_ERROR
    }
    MsgSender.sendMsg2Player(new ResponseVo("mahjongRoomService", "createRoomByEachUser", roomInfo.toJSONObject), userId)
    return 0
  }


  /**
    * 普通房
    *
    * @param userId
    * @param modeTotal
    * @param mode
    * @param multiple
    * @param gameNumber
    * @param personNumber
    * @param gameType
    * @param roomType
    * @param mustZimo
    * @param yipaoduoxiang
    * @param canChi
    * @param haveTing
    * @return
    */
  def createRoomByUser(userId: Long, modeTotal: String, mode: String, multiple: Int, gameNumber: Int, personNumber: Int,
                       gameType: String, roomType: String, mustZimo: Int, yipaoduoxiang: Boolean, canChi: Boolean,
                       haveTing: Boolean, clubId: String, clubRoomModel: String, clubMode: Int, goldRoomType: Int, goldRoomPermission: Int, showChat: Boolean, otherMode: Int): Int = {
    if (!isCanCreate(modeTotal, mode, "" + multiple)) {
      return ErrorCode.CANNOT_CREATE_ROOM_PARAMETER_IS_ERROR
    }
    val (code, roomInfo) = createRoom(userId, modeTotal, mode, multiple, gameNumber, personNumber, gameType,
      "0", true, roomType, mustZimo, yipaoduoxiang, canChi, haveTing,
      clubId, clubRoomModel, clubMode, goldRoomType, goldRoomPermission, showChat, otherMode)
    if (code != 0) {
      return code
    }
    MsgSender.sendMsg2Player(new ResponseVo("mahjongRoomService", "createRoomByUser", roomInfo.toJSONObject), userId)
    return 0
  }

  /**
    * 代建房
    *
    * @param userId
    * @param modeTotal
    * @param mode
    * @param multiple
    * @param gameNumber
    * @param personNumber
    * @param gameType
    * @param roomType
    * @param mustZimo
    * @param yipaoduoxiang
    * @param canChi
    * @param haveTing
    * @return
    */
  def createRoomButNotInRoom(userId: Long, modeTotal: String, mode: String, multiple: Int, gameNumber: Int, personNumber: Int,
                             gameType: String, roomType: String, mustZimo: Int, yipaoduoxiang: Boolean, canChi: Boolean, haveTing: Boolean,
                             clubId: String, clubRoomModel: String, clubMode: Int, goldRoomType: Int, goldRoomPermission: Int, showChat: Boolean, otherMode: Int): Int = {
    if (!isCanCreate(modeTotal, mode, "" + multiple)) {
      return ErrorCode.CANNOT_CREATE_ROOM_PARAMETER_IS_ERROR
    }
    //    val money: Double = RedisManager.getUserRedisService.getUserMoney(userId)
    //    val need = RoomInfo.getCreateMoney(gameType, gameNumber)
    //    if(money < need) {
    //      return ErrorCode.CANNOT_JOIN_ROOM_NO_MONEY
    //    }
    //    RedisManager.getUserRedisService.addUserMoney(userId, -need)
    //    if ("LQ" == gameType) {
    //      RedisManager.addGold(userId, need/10)
    //    }
    val (code, roomInfo) = createRoom(userId, modeTotal, mode, multiple, gameNumber, personNumber,
      gameType, "2", false, roomType, mustZimo, yipaoduoxiang, canChi, haveTing,
      clubId, clubRoomModel, clubMode, goldRoomType, goldRoomPermission, showChat, otherMode)
    if (code != 0) {
      return code
    }


    if (RedisManager.getUserRedisService.getUserMoney(userId) < roomInfo.getCreateNeedMoney) {
      RoomManager.removeRoom(roomInfo.getRoomId)
      return ErrorCode.CANNOT_JOIN_ROOM_NO_MONEY
    }
    //给代建房 开房者 扣钱
    roomInfo.spendMoney()


    val roomId = roomInfo.getRoomId
    val start: Long = System.currentTimeMillis
    val node: TimerNode = roomInfo.getDissolutionRoomTimerNode
    roomInfo.setPrepareRoomTimerNode(node)
    GameTimer.addTimerNode(node)
    MsgSender.sendMsg2Player(new ResponseVo("mahjongRoomService", "createRoomButNotInRoom", roomInfo.toJSONObject), userId)
    return 0
  }


  def joinGoldRoom(userId: Long, roomType: String, gameType: String, goldRoomType: Int): Int = {
    val rooms = RoomManager.getInstance().getNotFullRoom(gameType, goldRoomType)
    var room: Room = null
    var isAdd = false
    if (rooms.size() > 0) {
      room = rooms.get(0)
    }
    if (room == null) {
      isAdd = true
      room = new RoomInfoExtendGold()
      //获得一个默认房间
      room = room.getDefaultGoldRoomInstance(userId, roomType, gameType, goldRoomType)
    }
    val rtn: Int = room.joinRoom(userId, true)

    if (rtn != 0) {
      return rtn
    } else {
      if (isAdd) {
        val serverId: Int = SpringUtil.getBean(classOf[ServerConfig]).getServerId
        RoomManager.getInstance().addNotFullGoldRoom(room)
        RoomManager.addRoom(room.getRoomId, "" + serverId, room)
      }
      MsgSender.sendMsg2Player(new ResponseVo("mahjongRoomService", "joinGoldRoom", room.asInstanceOf[RoomInfo].toJSONObject), userId)
    }
    return 0

  }

}